File: backend/src/main/java/com/openclassrooms/mdd/service/user/UserService.java
package com.openclassrooms.mdd.service.user;

import org.springframework.security.core.userdetails.UserDetailsService;

import com.openclassrooms.mdd.dto.request.LoginRequest;
import com.openclassrooms.mdd.dto.request.RegisterRequest;
import com.openclassrooms.mdd.dto.request.UpdateProfileRequest;
import com.openclassrooms.mdd.dto.response.LoginResponse;
import com.openclassrooms.mdd.dto.response.UserProfileResponse;

public interface UserService extends UserDetailsService {

    /**
     * Registers a new user in the system.
     *
     * @param request the registration data (email, username, password)
     * @return the authentication token for the newly registered user
     */
    LoginResponse register(RegisterRequest request);

    /**
     * Authenticates a user with email/username and password.
     *
     * @param request the login credentials (identifier and password)
     * @return the authentication token for the logged-in user
     */
    LoginResponse login(LoginRequest request);

    /**
     * Retrieves the profile of the authenticated user.
     *
     * @param userEmail the email of the authenticated user (extracted from JWT)
     * @return the user's profile information including subscriptions
     */
    UserProfileResponse getUserProfile(String userEmail);

    /**
     * Updates the profile of the authenticated user.
     *
     * @param userEmail the email of the authenticated user (extracted from JWT)
     * @param request the update data (email, username, password - all optional)
     * @return the updated user profile
     */
    UserProfileResponse updateProfile(String userEmail, UpdateProfileRequest request);
}

File: frontend/src/app/core/services/user.service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { UserProfile } from '../models/user/user-profile.model';
import { UpdateProfileRequest } from '../models/user/update-profile-request.model';

@Injectable({
  providedIn: 'root',
})
export class UserService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/users`;

  /**
   * Get the current user's profile including subscriptions
   */
  getUserProfile(): Observable<UserProfile> {
    return this.http.get<UserProfile>(`${this.apiUrl}/me`);
  }

  /**
   * Update the current user's profile
   * @param updates Profile updates (email, username, password are optional)
   */
  updateProfile(updates: UpdateProfileRequest): Observable<UserProfile> {
    return this.http.put<UserProfile>(`${this.apiUrl}/me`, updates);
  }
}
